tests/ApiApprovalTests/GraphQL.Server.Transports.AspNetCore.approved.txt

```diff
@@ -1,55 +1,93 @@
+        protected GraphQL.Server.Transports.AspNetCore.GraphQLHttpMiddlewareOptions Options { get; }
+        protected virtual System.Threading.Tasks.ValueTask<System.Collections.Generic.IDictionary<string, object?>> BuildUserContextAsync(Microsoft.AspNetCore.Http.HttpContext context) { }
+        protected abstract System.Threading.Tasks.Task<GraphQL.ExecutionResult> ExecuteRequestAsync(Microsoft.AspNetCore.Http.HttpContext context, GraphQL.Transport.GraphQLRequest? request, System.IServiceProvider serviceProvider, System.Collections.Generic.IDictionary<string, object?> userContext);
+        protected abstract System.Threading.Tasks.Task<GraphQL.ExecutionResult> ExecuteScopedRequestAsync(Microsoft.AspNetCore.Http.HttpContext context, GraphQL.Transport.GraphQLRequest request, System.Collections.Generic.IDictionary<string, object?> userContext);
+        protected virtual System.Threading.Tasks.ValueTask<bool> HandleAuthorizeAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.ValueTask<bool> HandleAuthorizeWebSocketConnectionAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.Task HandleBatchRequestAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next, System.Collections.Generic.IList<GraphQL.Transport.GraphQLRequest?> gqlRequests) { }
+        protected virtual System.Threading.Tasks.Task HandleBatchedRequestsNotSupportedAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.Task HandleContentTypeCouldNotBeParsedErrorAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.ValueTask<bool> HandleDeserializationErrorAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next, System.Exception exception) { }
+        protected virtual System.Threading.Tasks.Task HandleInvalidContentTypeErrorAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.Task HandleInvalidHttpMethodErrorAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.Task HandleNotAuthenticatedAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next) { }
+        protected virtual System.Threading.Tasks.Task HandleNotAuthorizedPolicyAsync(Microsoft.AspNetCore.Http.HttpContext context, Microsoft.AspNetCore.Http.RequestDelegate next, Microsoft.AspNetCore.Authorization.AuthorizationResult authorizationResult) { }
```

Comment:
Too many `HandleXXX` methods for different error conditions. I would try to extract them into dedicated class out of middleware to keep middleware code focused - 700 lines of code now.

Reply:
I'd leave them, but perhaps we add an enum and then consolidate the methods together -- for all the methods with the same signature.

```cs
protected virtual Task HandleError(HttpContext context, RequestDelegate next, ErrorCondition errorCondition);

enum ErrorCondition
{
    InvalidContentType,
    InvalidHttpMethod,
    NotAuthenticated,
}
```

Or something.  Again, I would prefer not to change it, but am open to exploring another option here.  I want to be sure that we can somehow override behavior to change an error message (or behavior) without re-implementing any more code than necessary.  Such as the following for the above example:

```cs
protected override Task HandleError(HttpContext context, RequestDelegate next, ErrorCondition errorCondition)
{
    if (errorCondition == ErrorCondition.InvalidContentType)
    {
        // do something
    }
    else
        return base.HandleError(context, next, errorCondition);
}
```

Reply:
> 700 lines

While true, there are a significant number of lines dedicated to comments, a large number necessary for WebSocket support, and a derived class that could be moved to another file.  The main `InvokeAsync` method is only 150 lines of code, with well-documented methods (via naming and xml comments) making all of the code easy to follow and understand.

In my company, I have a rule of thumb that methods "cannot be more than two pages of code".  This encourages splitting up functionality into understandably-sized pieces, and I have tried to adhere to that principle here as well.  Here we have about 2.5-3 pages of code in the InvokeAsync method.  If anything, we could probably split up the code further -- for example splitting the code to read the POST body content to another method.  This has an added benefit of allowing derived classes to implement another content type, if desired.  But even as-is, I think the code is pretty easy to understand.

Reply:
#795

Reply:
I make my proposal based on how I do review. It's more comfortable to go through small files and mark them reviewed. If the files are large and I do not look at them at a time, then returning to review I have to completely read them again to remember what I did.